Shared IPC type contract between renderer and main process

niik opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

## Description
<!--
A summary of the changes made along with any other information that would be helpful to a reviewer such as potential tradeoffs or alternative approaches you considered.
-->

This relates to #13689 which we discussed this synchronously today. Our current IPC approach relies on copy-pasted code between the renderer and the main process where we hope that the channel names and the type signature match between the two.

This attempts to reduce that brittle approach by sharing a common type contract between the renderer and the main process.

### Screenshots

<!--
If this PR touches the UI layer of the app, please include screenshots or animated gifs to show the changes.
-->

## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes:

niik committed:
Create strongly typed IPC methods

niik committed:
Use IPC proxies in main-process-proxy

niik committed:
:art: cleanup

niik committed:
Use typed IPC methods in main process

niik committed:
Leverage type helpers in renderer logger

niik committed:
Leverage type helpers in app shell

niik committed:
How did this ever work?

niik committed:
Use type helpers in AppWindow

tidy-dev commented:
Approach looks great to me! Thank you for doing this typing wizardry to make things so much prettier!

I like the names on `sendProxy`, and `invokeProxy`, since they directly correlate with the IPC `send` and `invoke` methods and because of that the `onIPCMainRequest` feel off.. part of me wants it to say `onIPCMainHandleRequest` or have `onProxy` and `handleProxy` for consistency....and maybe all of those could have `IPC` in them. `sendIPCProxy`, `invokeIPCProxy`, `onIPCProxy`, `handleIPCProxy`. Just my random thoughts. :) 

What do you all think about also at this time adding a IPCChannelName enum since we have to use the channel names in three places (renderer, main, and channel definition)?

Also, in the electron upgrade, I refactored the autoUpdater logic to use the main process proxy and added "on" methods there that took a function as the param to move all the ipc channel references to the proxy. (In the #13689, you can see what I am talking about in the `update-store.ts` changes). Thoughts on adding this to the approach?

niik commented:
> I like the names on `sendProxy`, and `invokeProxy`, since they directly correlate with the IPC `send` and `invoke` methods and because of that the `onIPCMainRequest` feel off.. part of me wants it to say `onIPCMainHandleRequest` or have `onProxy` and `handleProxy` for consistency....and maybe all of those could have `IPC` in them. `sendIPCProxy`, `invokeIPCProxy`, `onIPCProxy`, `handleIPCProxy`. Just my random thoughts. :)

I love your random thoughts. I'll give this some more thought. Maybe we could try duplicating the ipcXYZ modules. Something like 

```ts
import * as ipcMain from 'ipc-main.ts'

ipcMain.on(...)
ipcMain.handle(...)
```

> What do you all think about also at this time adding a IPCChannelName enum since we have to use the channel names in three places (renderer, main, and channel definition)?

I'm not sure I see what an enum would give us that the current approach doesn't. The strings are effectively constants already as you can't pass in something that's not defined in the channel types and you already get the same level of autocompletion that you'd get by using an enum. They only thing I can think of is that they would let us put docstrings on the enum values but I doubt we would. As for the duplication we would still have to use the channel name _enum value_ in three places (renderer, main, and channel definition)?

> Also, in the electron upgrade, I refactored the autoUpdater logic to use the main process proxy and added "on" methods there that took a function as the param to move all the ipc channel references to the proxy. (In the #13689, you can see what I am talking about in the `update-store.ts` changes). Thoughts on adding this to the approach?

Yeah definitely, adding a type safe `.on` and `.handle` for the renderer as well as the main process is certainly on my list!

niik committed:
This was a supremely placed joke

This ancient snipped relied on the fact that ipcMain is a generic EventEmitter (even though you'd never use it that way) and emitted an event on it which was easily confused as an event from the renderer but since it wasn't it had a different signature.

tidy-dev commented:
> I'm not sure I see what an enum would give us 

😸  Being able to say `IPCChannelName.` and visual studio code will help me finish it, so I don't have to worry about my ineptness of copy and paste. But like you said the build:dev yells at me when I do this, so either way is good.

tidy-dev commented:
> Maybe we could try duplicating the ipcXYZ modules. 

I like the sound of that.

niik commented:
> Being able to say `IPCChannelName.` and visual studio code will help me finish it, so I don't have to worry about my ineptness of copy and paste. But like you said the build:dev yells at me when I do this, so either way is good.

In the current case you only have to type a single `'` and you'll get the exact same level of autocompletion though.

https://user-images.githubusercontent.com/634063/150398186-ba215de0-aabd-4b8f-9336-7e23eb8150a4.mov

tidy-dev commented:
Ah yeah! I forgot that the new typing thing will give us that too! This things is so cool!

sergiou87 commented:
Sorry I didn't have much time to look at this yesterday, but I like it overall, and the idea of "replacing"  the IPC objects sounds great! 😄

niik committed:
Try matching the current ipcMain API

niik committed:
Match ipcRenderer API

niik committed:
Use typed IPC in crash app

niik committed:
Convert more IPC events to typed ipc

niik committed:
Forgot one

niik committed:
Add lint rules to prevent us from using ipcRenderer

niik committed:
Remove more ipcRenderer uses

niik committed:
:fire: unused method

niik committed:
Prevent importing ipcMain straight from electron

niik committed:
Add a rule for not accidentally using webContents.send directly

niik committed:
Strongly typed web contents ipc

niik committed:
Find even more of these buggers

niik committed:
Dunno why that can't resolve

niik committed:
Merge branch 'development' into typed-ipc

niik commented:
Alright,this is ready for a little lookiloo again. I've tried the approach of matching the API (albeit only a subset of it) and I dig it.

niik committed:
Use the type-safe ipcRenderer in main-process-proxy

tidy-dev commented on app/src/lib/logging/renderer/install.ts:
```diff
@@ -1,20 +1,17 @@
+import { sendProxy } from '../../../ui/main-process-proxy'
 
 const g = global as any
+const ipcLog = sendProxy('log')
```
Nit: For consistency with other calls like `moveToTrash`, seems like this could be in the `main-process-proxy` and import `ipcLog` from there instead of `sendProxy`.

tidy-dev reviewed (approved):
✨  Looks great! I pulled and tested it out. Also been working with it on my 16 upgrade branch when refactor remote references and so far it has worked and been fantastic.

One little nit, but feel free to ignore.

sergiou87 commented on .eslintrc.yml:
```diff
@@ -149,6 +150,29 @@ rules:
     - error
     - global
   no-buffer-constructor: error
+  no-restricted-imports:
+    - error
+    - paths:
+        - name: electron
+          importNames: ['ipcRenderer']
+          message:
+            "Please use 'import * as ipcRenderer' from 'ipc-renderer' instead to
+            get strongly typed IPC methods."
+        - name: electron/renderer
+          importNames: ['ipcRenderer']
+          message:
+            "Please use 'import * as ipcRenderer' from 'ipc-renderer' instead to
+            get strongly typed IPC methods."
+        - name: electron
+          importNames: ['ipcMain']
+          message:
+            "Please use 'import * as ipcMain' from 'ipc-main' instead to get
+            strongly typed IPC methods."
+        - name: electron/main
+          importNames: ['ipcMain']
+          message:
+            "Please use 'import * as ipcMain' from 'ipc-main' instead to get
+            strongly typed IPC methods."
```
TIL about `no-restricted-imports` 😍

sergiou87 reviewed (approved):
I love it!! Awesome work 😍
